feat: input, textarea, auto-complete

yqrashawn opened the pull request:
## Checklist

- [x] Fix linting errors
- [x] Tests have been added / updated (or snapshots)

## Change information

### input

1. add variant, stats -> color, misc style
2. remove some unused style from doc

### textarea

1. add variant, stats -> color, misc style
2. remove some unused style from doc
3. add word count ui, `maxLength` prop automatically enables the counter ui

### auto-complete

1. add variant, stats -> color, misc style
2. remove some unused style from doc
3. add fuzzy-match style to none custom auto-complete options

yqrashawn committed:
feat(input): add onHover style

yqrashawn committed:
feat(input): change default colorbloc

yqrashawn committed:
feat(input): handle hover with js, add hoverColor

yqrashawn committed:
feat(input): disabled style

yqrashawn committed:
feat(input): change size default style

yqrashawn committed:
docs(input): demonstrate size for input with icon

yqrashawn committed:
docs(input): use icon from designer

yqrashawn committed:
feat(input): add primary input

yqrashawn committed:
fix(input): try fix input line-height problem

yqrashawn committed:
fix(input): disabled input don't have hover border effect

yqrashawn committed:
fix(input): use cGray5 for input placeholder

yqrashawn committed:
test(input): snapshots

yqrashawn committed:
feat(input): change input clear icon

yqrashawn committed:
fix(input): firefox line-height bug

yqrashawn committed:
fix(input): always synchronize external value and ignore changes (#336)

* fix(input): always synchronize external value and ignore changes

* feat(input): support imperative API to update

(cherry picked from commit 7dabdac557e0b430b6274219d0d398c998de74a8)

yqrashawn committed:
fix: add default export for test icon

yqrashawn committed:
fix: remove test icons

yqrashawn committed:
fix(input): input icon default to 14px

yqrashawn committed:
feat(input): add solid input

yqrashawn committed:
fix(input): typo visibale -> visible

yqrashawn committed:
fix(textarea): hoverBorder -> hoverBorderColor

yqrashawn committed:
fix(input): solid secondary bg color

yqrashawn committed:
feat(textarea): add opacity to disabled textarea

yqrashawn committed:
feat(textarea): separate hover and focus

yqrashawn committed:
fix(textarea): imperative api same as input

yqrashawn committed:
fix(input): more consistent style impl

yqrashawn committed:
feat(input): adjust color

yqrashawn committed:
feat(textarea): solid style

yqrashawn committed:
fix(input): no style when hovering disabled inputs

yqrashawn committed:
feat(textarea): add counter to textarea

yqrashawn committed:
feat(input): no need to have secondary input style

yqrashawn committed:
docs(textarea): textarea doc

yqrashawn committed:
docs(input): update doc

yqrashawn committed:
fix(textarea): brwoser auto fill style

yqrashawn committed:
fix(input): input textarea status type use InputTypes

yqrashawn committed:
feat(input): solid input label style

yqrashawn committed:
test(input): test cover more code

yqrashawn committed:
docs(input): input doc style

yqrashawn committed:
test(input): input, textarea test coverage

yqrashawn committed:
fix(textarea): disabled textarea has no hover style

yqrashawn committed:
fix(textarea): counter and place holder style

yqrashawn committed:
feat(auto-complete): lined style

yqrashawn committed:
fix(auto-complete): change input and dropdown styles at the same time

yqrashawn committed:
doc(auto-complete): add console.log in creatable

yqrashawn committed:
fix(input): new color

yqrashawn committed:
test(auto-complete): test solid auto complete

yqrashawn committed:
fix(input): missing useMemo dependency

yqrashawn committed:
feat(auto-complete): solid style

yqrashawn committed:
docs(input): use size prop value as placeholder

yqrashawn committed:
fix(auto-complete): item hover color

yqrashawn committed:
fix(auto-complete): dropdown timer

yqrashawn committed:
feat(auto-complete): add fuzzy-match

yqrashawn committed:
fix(input): solid? -> variant

yqrashawn committed:
docs(auto-complete): add icon left example

yqrashawn committed:
fix(input): remove error hover

yqrashawn committed:
fix(input): status -> color

yqrashawn committed:
docs(textarea): add counter doc

yqrashawn committed:
chore(codeowner): add codeowner for textarea and auto-complete

yqrashawn committed:
refactor(input): update snapshots to change pt to px

yqrashawn commented on components/auto-complete/auto-complete-searching.tsx:
```diff
@@ -32,7 +32,6 @@ const AutoCompleteSearch: React.FC<React.PropsWithChildren<AutoCompleteSearchPro
           padding: ${theme.layout.gapHalf};
           line-height: 1;
-          background-color: ${theme.palette.background};
           color: ${theme.palette.accents_5};
```
forget to change the color here

yqrashawn commented on components/input/input-label.tsx:
```diff
@@ -29,26 +40,26 @@ const InputLabel: React.FC<React.PropsWithChildren<InputLabel>> = ({
           background-color: ${theme.palette.accents_1};
```
forget to change the color here

yqrashawn reviewed (commented):
forget to change some original color

yqrashawn commented on components/input/input-icon.tsx:
```diff
@@ -6,15 +6,24 @@ export interface InputIconProps {
   const width = useMemo(() => {
-    return `calc(${ratio} * ${theme.layout.gap} * .42)`
+    return `calc(${ratio} * ${theme.layout.gap} * .35)`
   }, [theme.layout.gap, ratio])
   const padding = useMemo(() => {
-    return `calc(${ratio} * ${theme.layout.gap} * .3)`
+    return `calc(${ratio} * ${theme.layout.gap} * .4)`
```
check these style after pt changed to px

yqrashawn commented on components/input/styles.ts:
```diff
@@ -1,28 +1,34 @@
-      heightRatio: '1.313',
-      fontSize: '.75rem',
+      heightRatio: '1.125',
+      fontSize: '0.5714rem',
+      margin: '0 1.1429rem',
     },
     small: {
       heightRatio: '1.5',
-      fontSize: '.75rem',
+      fontSize: '0.8571rem',
+      margin: '0 1.1429rem',
```
check these styles after pt changed to px

jnoodle commented on components/auto-complete/auto-complete-dropdown.tsx:
```diff
@@ -21,35 +24,61 @@ export type AutoCompleteDropdownProps = Props & typeof defaultProps & NativeAttr
+      leaveTime={value === '' ? 0 : 60}>
       <div
-        className={`auto-complete-dropdown ${className}`}
+        className={`auto-complete-dropdown ${className} ${isSolid ? 'solid' : 'line'}`}
```
Should increase custom className's priority

yqrashawn replied:
You mean put the `${className}` after the solid/line?

jnoodle replied:
yes

yqrashawn replied:
Done

jnoodle commented on components/textarea/counter.tsx:
```diff
@@ -0,0 +1,44 @@
+
+const Counter: React.FC<React.PropsWithChildren<CounterProps>> = ({ count, maxLength }) => {
+  const theme = useTheme()
+  const hasLimit = useMemo(() => Number.isInteger(maxLength) && (maxLength as number) > 0, [
```
has a bug: try to set maxLength to 1000000000000000000000

yqrashawn replied:
done

jnoodle commented on components/shared/fuzzy-match.tsx:
```diff
@@ -0,0 +1,94 @@
+  ]
+
+  if (pattern === 'fuzzy') {
+    query = query.trim().replace(/\ /, '.*')
```
is add `/g` better?

yqrashawn replied:
done

jnoodle commented on pages/zh-cn/components/auto-complete.mdx:
```diff
@@ -228,42 +309,21 @@ export const meta = {
   title="可清除的"
   desc="在输入框内添加一个用作清除文字的按钮。"
-  scope={{ AutoComplete }}
+  scope={{ AutoComplete, Spacer }}
```
Should add `SearchIcon`

yqrashawn replied:
done

jnoodle commented on components/shared/fuzzy-match.tsx:
```diff
@@ -0,0 +1,94 @@
+  let beforeMatched = ''
+  let matched = ''
+  let afterMatched = ''
+  if (pattern === 'fuzzy' || pattern === 'regex') {
```
bug: set pattern to `regex`, eg: `<AutoComplete placeholder="Enter here" options={options} pattern="regex" />` 

then input `?`, throw `SyntaxError: Invalid regular expression: /?/: Nothing to repeat`

jnoodle replied:
the same goes for direct input, try `?`, `[` ...

yqrashawn replied:
Done. I removed the `pattern` prop.

yqrashawn committed:
fix(textarea): maxLength bug

yqrashawn committed:
fix(fuzzy-match): remove pattern, escape regex

yqrashawn committed:
docs(auto-complete): SearchIcon in playground scope

0x74616e67 commented on components/auto-complete/auto-complete-dropdown.tsx:
```diff
@@ -21,35 +24,61 @@ export type AutoCompleteDropdownProps = Props & typeof defaultProps & NativeAttr
   dropdownStyle,
   disableMatchWidth,
 }) => {
+  const isSolid = useMemo(() => variant === 'solid', [variant])
```
I suggest not to use useMemo here, because equal condition is not expensive calculations, and the dependance array list will make equal check every time too.

yqrashawn replied:
done

yqrashawn committed:
fix(input): remove accents color

yqrashawn committed:
fix(input): resolve review

1. change styles affected by pt->px change
2. remove useless useMemo
3. remove original background/forground/accents_* color
4. format mdx docs
